### ISSUE TITLE
Generators Always Match in Palace Rooms, Even When Flag Disabled

### ISSUE THREAD

**Issue by MiR4i** [Ellendar/Z2Randomizer#101]
As the title states, when the "Generators Always Match" flag is disabled, only one generator type will ever appear in a palace room.  However, the overworld encounters, including when inside caves, seems to respect this flag.

There may be further issues with generator enemies in palace rooms, but I'll wait to see if fixing this issue clears up that issue before reporting it.

This issue is not specific to any particular palace room, but here are a few examples when the flag is disabled:

![image](https://github.com/Ellendar/Z2Randomizer/assets/6069346/2404b34a-0734-443f-9489-c0e9eec934d3)

**Comment by Ellendar** [comment]
I looked at it a bit and there isn't anythig obviously wrong. Without knowing the room/palace group/flags, it's probably impossible for me to find the issue without at least a couple seeds on the same flagset that have the same room.

**Comment by Ellendar** [comment]
I did find AN error, but it wouldn't describe all your issues. Basically the enemy randomization logic takes each enemy and says "is this a small enemy? if so make it a new random small enemy" and repeat with each other kind of enemy. The problem is that the list of enemies you can randomize INTO, and the list of enemies you can randomize FROM are the same, and they're separate in each palace group. So if you have a room where the vanilla enemy is only found in palace 125 like the palace 125 version of a sin head generator (0x0F), if that enemy is in palace 346, it doesn't see it as being any kind of enemy, as it's not a valid enemy type in that palace group, so it just won't randomize. The safety catch to avoid putting 0x0F in eastern will then convert it to a normal sin generator, so effectively any room that was originally created to have sin wave heads in palace 125, will never randomize if it gets put into palace 346 (but will still work fine in 125)

A quick fix is to just check the list of all enemies from all palaces to determine what group an enemy is, and then pick the shuffled enemy from the list of valid enemies.

**Comment by MiR4i** [comment]
Flagset: `RAAN6AAFesXtN1Tdt0Y$8v4cYf4XT7AALh`
Seed: `330018372`
Rooms:
```
P1-07
P2-13
P4-03
P4-07
P4-17
P5-14
P5-20
P6-15
P6-18
P6-19
P6-22
P7-06
P7-07
P7-22
P7-25
P7-32
```
The above are just the rooms that are on the comparison chart below, and there are a handful more in this particular seed that are not listed because I could not find a seed in 4.0.4 that had the same room to put into the chart.  However, ***I will add that this issue is not specific to just these rooms, but every palace room that contains a generator***.

Comparison Chart: https://z2r.gg/content/images/matching_generators_bug.png

Furthermore, this is not a cherry-picked seed, flagset, or set of rooms, as I've played with numerous flagsets and seeds where I've disabled matching generators and ***they all exhibit the same exact behavior as shown in the above chart***.  If you can provide a flagset and seed that contains palaces with rooms that have non-matching generators, I would, in all honesty, like to look at it.

**Comment by Ellendar** [comment]
Part of the reason I asked you to hold off until I pushed the change is because I wanted to see how much of the effect was the enemy shuffle bug I found. The other half was that the change alters palace generation, so I can't reliably tell whether what it generates is the same as what you're referring to.

**Comment by MiR4i** [comment]
This has been fixed in 4.2.5 and can be closed.
